tests/environment.py

```diff
@@ -0,0 +1,299 @@
+    checking its status.
+    '''
+    def __init__(self, agent: ALEntity, mac: str):
+        self.agent = agent
+        agent.radios.append(self)
+        self.mac = mac
+        self.vaps = []
+
+    def wait_for_log(self, regex: str, timeout: float) -> bool:
+        '''Poll the radio's logfile until it contains "regex" or times out.'''
+        raise NotImplementedError("wait_for_log is not implemented in abstract class Radio")
+
+
+class Station:
+    '''Placeholder for a wireless (fronthaul) station.
+
+    Unlike the other classes, this is not an abstract class. Instead, it is a placeholder that
+    represents a station. Handling the station is actually done through the VirtualAP concrete
+    implementation.
+    '''
+    def __init__(self, mac: str):
+        self.mac = mac
+
+    __mac_base = 0
+
+    @staticmethod
+    def create():
+        '''Generate a Station placeholder with a random MAC address.'''
+        mac = '51:a1:10:20:{:02x}:{:02x}'.format(int(Station.__mac_base / 256),
+                                                 Station.__mac_base % 256)
+        Station.__mac_base += 1
+        if Station.__mac_base > 256*256:
+            Station.__mac_base = 0
+        return Station(mac)
+
+
+class VirtualAP:
+    '''Abstract representation of a VAP on a MultiAP Radio.'''
+    def __init__(self, radio: Radio, bssid: str):
+        self.radio = radio
+        radio.vaps.append(self)
+        self.bssid = bssid
+
+    def associate(self, sta: Station) -> bool:
+        '''Associate "sta" with this VAP.'''
+        raise NotImplementedError("associate is not implemented in abstract class VirtualAP")
+
+    def disassociate(self, sta: Station) -> bool:
+        '''Disassociate "sta" from this VAP.'''
+        raise NotImplementedError("disassociate is not implemented in abstract class VirtualAP")
+
+
+# The following variables are initialized as None, and have to be set when a concrete test
+# environment is started.
+wired_sniffer = None
+controller = None
+agents = []
+
+
+def beerocks_cli_command(command: str) -> bytes:
+    '''Execute `command` beerocks_cli command on the controller and return its output.'''
+    debug("Send CLI command " + command)
+    res = controller.prplmesh_command("bin/beerocks_cli", "-c", command)
+    debug("  Response: " + res.decode('utf-8', errors='replace').strip())
+    return res
+
+
+# Concrete implementation with docker
+
+rootdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+installdir = os.path.join(rootdir, 'build', 'install')
+
+
+def _docker_wait_for_log(container: str, program: str, regex: str, timeout: float) -> bool:
+    logfilename = os.path.join(rootdir, 'logs', container, 'beerocks_{}.log'.format(program))
+    deadline = time.monotonic() + timeout
+    try:
+        while True:
+            with open(logfilename) as logfile:
+                for line in logfile.readlines():
+                    if re.search(regex, line):
```

Comment:
This implementation does not include a case when logfile contains 2 or more same words. It just takes the first found word.

Reply:
Well, that feature is something that you added, no?